### PR TITLE
Fixes #37

### LINKED ISSUE #37: Bug with loss parsing
https://github.com/ThrunGroup/BanditPAM/blob/main/src/main.cpp#L40-L42 contains a bug; it assumes that the single character after the `L` is the norm; this creates an error, for example, when the loss is `L15`, or if the name of a loss function happens to start with `L`.

Instead, we should check specifically for the form `L\d*` and get all characters except the first in that case.

### PR DESCRIPTION
**Pull request by mailology**
Fix bug in loss parsing by checking the format of the input loss string. When user enters loss in the form of `"L\d*"`, the whole numeric portion will be read into the variable `loss` and is treated as the `p` value in Lp norm. Loss function with the name starting with `L` but not in the form of `"L\d*"` will remain intact (but further implementation needs to be done in `kmedoids_ucb.cpp`).

For example:
1. `./BanditPAM -f ../../data/MNIST-1k.csv -k 10 -v 1 -l L15` will give `Medoids: 194,319,558,60,133,896,448,475,251,354`. 
2. `./BanditPAM -f ../../data/MNIST-1k.csv -k 10 -v 1  ` will give the expected result `Medoids: 694,800,306,714,324,959,737,527,168,251`.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -56,7 +57,7 @@ int main(int argc, char* argv[])
             // type of loss/distance function to use
             case 'l':
                 loss = optarg;
-                if (loss.at(0) == 'L') {
+                if (std::regex_match(loss, std::regex("L\\d*"))) {
```
Comment (motiwari) [inline src/main.cpp:60]: This is great! But also needs to be changed here: https://github.com/ThrunGroup/BanditPAM/blob/main/src/kmedoids_ucb.cpp#L110

**Review by motiwari — CHANGES_REQUESTED** [review]

File: src/main.cpp (hunk 1 above)
Comment (mailology) [reply src/main.cpp:60]: On https://github.com/ThrunGroup/BanditPAM/blob/main/src/kmedoids_ucb.cpp#L110, the checking for loss starting with `L` here becomes redundant since the Lp norm is already checked in the `main.cpp` and the p is parsed and saved in `loss` variable.

File: src/kmedoids_ucb.cpp [hunk 2]
```diff
@@ -107,9 +107,6 @@ int KMedoids::getSteps() {
  *  @param loss Loss function to be used e.g. L2
  */
 void KMedoids::setLossFn(std::string loss) {
-  if (loss.at(0) == 'L') {
```
[the remark this answers is not in the document]
Comment (mailology) [reply src/kmedoids_ucb.cpp:110]: https://github.com/ThrunGroup/BanditPAM/blob/f28d3096bf65fcf026d00f2ba3784f995cb13734/src/main.cpp#L58-L62
We read the loss in the `main.cpp`. The condition here checks if the loss starts with `L` following with digits. For example if we enter argument `-l L15`, the condition is satisfied and we assign `loss = 15`. On the other hand if we enter argument which does not satisfy the condition, say, `-l L15p`, then we assign `loss = L15p`. 

Then for the case `loss = 15` we set the `Lp` norm here:
https://github.com/ThrunGroup/BanditPAM/blob/f28d3096bf65fcf026d00f2ba3784f995cb13734/src/kmedoids_ucb.cpp#L116-L118

For the case where `loss = L15p`, we may assume this loss is not recognized and handled by the following
https://github.com/ThrunGroup/BanditPAM/blob/9d1e6c86a915ffbee877f66f447ecd917c1aeb96/src/kmedoids_ucb.cpp#L125-L127

Here are some examples I have tried:
```
1. ./BanditPAM -f ../../data/MNIST-1k.csv -k 10 -v 1 -l L2: Medoids: 694,800,306,714,324,959,737,527,168,251
2. ./BanditPAM -f ../../data/MNIST-1k.csv -k 10 -v 1 -l L15: Medoids: 194,319,558,60,133,896,448,475,251,354
3. ./BanditPAM -f ../../data/MNIST-1k.csv -k 10 -v 1 -l manhattan: Medoids: 737,246,527,168,714,306,838,324,959,800
4. ./BanditPAM -f ../../data/MNIST-1k.csv -k 10 -v 1 -l cos: Medoids: 567,182,839,340,385,979,915,758,609,649
5. ./BanditPAM -f ../../data/MNIST-1k.csv -k 10 -v 1 -l L15p: libc++abi.dylib: terminating with uncaught exception of type char const*
```

The last one does not throw the expected message because there is no `std::invalid_argument()` to handle the message and `catch()` for the exception, I will push a new code to fix it.

[the remark this answers is not in the document]
Comment (motiwari) [reply src/kmedoids_ucb.cpp:110]: Ah, great! The second code block was what I missed. Could you also delete the comments right after line 118 in kmedoids_ucb.cpp?

New error message sounds good as well

[the remark this answers is not in the document]
Comment (mailology) [reply src/kmedoids_ucb.cpp:110]: Cool! Then maybe I just keep that error message and remove the comments right after line 118.

[the remark this answers is not in the document]
Comment (motiwari) [reply src/kmedoids_ucb.cpp:110]: Actually I think a better error message would still help users

[the remark this answers is not in the document]
Comment (mailology) [reply src/kmedoids_ucb.cpp:110]: Agree. I have edited the code to throw the error messages. Now we will have proper error message for case like `loss = L15p`:
```
./BanditPAM -f ../../data/MNIST-1k.csv -k 10 -v 1 -l L15p:  error: unrecognized loss function
```

Also, I have merged with the updated main branch.

File: src/kmedoids_ucb.cpp [hunk 3]
```diff
@@ -107,24 +107,22 @@ int KMedoids::getSteps() {
...
-  } else if (std::isdigit(loss.at(0))) {
-      lossFn = &KMedoids::LP;
-      lp     = atoi(loss.c_str());
-      //std::stringstream st(loss);
-      //st >> lp;
-      //lp     = std::stoi(loss);
-  } else {
-      throw "unrecognized loss function";
-  }
+  try {
```
Comment (motiwari) [inline src/kmedoids_ucb.cpp:110]: Sorry to backtrack on this, but I still think we need the parsing of the loss in this function, for the same reason we include the other `if` statements below:

This is because we should not assume that the loss parsing is done by the `main` function in `main.cpp`. In fact, when called by Python, we do not call `main` -- we actually create a binding that calls `setLossFn` directly. For that reason, we should include the loss-parsing code *here*, not in `main.cpp`.

Furthermore, to avoid duplicate code, we should have `main` call *this* function to set the loss function. Right now, [this](https://github.com/ThrunGroup/BanditPAM/blob/03f93356830ee5ed2dfc4a905edf349619faa76e/src/main.cpp#L60-L63) only works for numeric loss functions and doesn't support `manhattan` etc. This will also be fixed once we have `main` call `setLossFn`

**Review by motiwari — CHANGES_REQUESTED** [review]

File: src/kmedoids_ucb.cpp (hunk 3 above)
Comment (mailology) [reply src/kmedoids_ucb.cpp:114]: Thanks @motiwari for this important point on Python! I have added back the loss parsing in `setLossFn` and retested all the cases in both C++ and Python:
```
1. ./BanditPAM -f ../../data/MNIST-1k.csv -k 10 -v 1 -l L2: Medoids: 694,800,306,714,324,959,737,527,168,251
2. ./BanditPAM -f ../../data/MNIST-1k.csv -k 10 -v 1 -l L15: Medoids: 194,319,558,60,133,896,448,475,251,354
3. ./BanditPAM -f ../../data/MNIST-1k.csv -k 10 -v 1 -l manhattan: Medoids: 737,246,527,168,714,306,838,324,959,800
4. ./BanditPAM -f ../../data/MNIST-1k.csv -k 10 -v 1 -l cos: Medoids: 567,182,839,340,385,979,915,758,609,649
5. ./BanditPAM -f ../../data/MNIST-1k.csv -k 10 -v 1 -l L15p: error: unrecognized loss function
```
Both of them give expected results for all cases. For duplication of code, maybe we could open a new issue and fix it there?

Comment (motiwari) [reply src/kmedoids_ucb.cpp:114]: Sounds good! @mailology could you file another issue for the duplicate code?

**Review by motiwari — APPROVED** [review]
